module/os/linux/zfs/zpl_file.c

```diff
@@ -998,6 +998,118 @@ zpl_ioctl_setxattr(struct file *filp, void __user *arg)
 	return (err);
 }
 
+/*
+ * Expose Additional File Level Attributes of ZFS.
+ */
+static int
+zpl_ioctl_getdosflags(struct file *filp, void __user *arg)
+{
+	struct inode *ip = file_inode(filp);
+	uint64_t dosflags = ITOZ(ip)->z_pflags;
+	dosflags &= ZFS_DOS_FL_USER_VISIBLE;
+	int err = copy_to_user(arg, &dosflags, sizeof (dosflags));
+
+	return (err);
+}
+
+static int
+__zpl_ioctl_setdosflags(struct inode *ip, uint64_t ioctl_flags, xvattr_t *xva)
+{
+	uint64_t zfs_flags = ITOZ(ip)->z_pflags;
+	xoptattr_t *xoap;
+
+	if (ioctl_flags & (~ZFS_DOS_FL_USER_VISIBLE))
+		return (-EOPNOTSUPP);
+
+	if ((fchange(ioctl_flags, zfs_flags, ZFS_IMMUTABLE, ZFS_IMMUTABLE) ||
+	    fchange(ioctl_flags, zfs_flags, ZFS_APPENDONLY, ZFS_APPENDONLY)) &&
+	    !capable(CAP_LINUX_IMMUTABLE))
+		return (-EPERM);
+
+	if (!zpl_inode_owner_or_capable(kcred->user_ns, ip))
+		return (-EACCES);
+
+	xva_init(xva);
+	xoap = xva_getxoptattr(xva);
+
+	XVA_SET_REQ(xva, XAT_IMMUTABLE);
+	if (ioctl_flags & ZFS_IMMUTABLE)
+		xoap->xoa_immutable = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_APPENDONLY);
+	if (ioctl_flags & ZFS_APPENDONLY)
+		xoap->xoa_appendonly = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_NODUMP);
+	if (ioctl_flags & ZFS_NODUMP)
+		xoap->xoa_nodump = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_READONLY);
+	if (ioctl_flags & ZFS_READONLY)
+		xoap->xoa_readonly = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_HIDDEN);
+	if (ioctl_flags & ZFS_HIDDEN)
+		xoap->xoa_hidden = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_SYSTEM);
+	if (ioctl_flags & ZFS_SYSTEM)
+		xoap->xoa_system = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_ARCHIVE);
+	if (ioctl_flags & ZFS_ARCHIVE)
+		xoap->xoa_archive = B_TRUE;
+
```

Comment:
@behlendorf  The original motivation for setting XAT_* attributes this way was taken from __zpl_ioctl_setflags(). But as @amotin pointed out, this could be an issue if the underlying implementation in zfs_setattr() changes. Perhaps it would be a good idea to generalize how FreeBSD is setting XAT_* attributes for Linux as well in __zpl_ioctl_setdosflags() and possibly __zpl_ioctl_setflags() ?

Reply:
My preference would be to keep the attribute handling consistent so can you make the same change to `__zpl_ioctl_setflags`.

Reply:
Updated the attribute handling for `__zpl_ioctl_setflags` as done in `__zpl_ioctl_setdosflags`.

Reply:
Hmm, this seems to have changed the expected user behavior and we have a few related test failures.

```sh
Tests with results other than PASS that are unexpected:
    FAIL projectquota/projecttree_001_pos (expected PASS)
    FAIL projectquota/projecttree_002_pos (expected PASS)
```

http://build.zfsonlinux.org/builders/CentOS%20Stream%208%20x86_64%20%28TEST%29/builds/4311/steps/shell_4/logs/summary

Reply:
Working on debugging it.